File: src/primitives/address.ts
/**
 * Address Types and Utilities
 *
 * Ethereum address (20 bytes) with type-safe operations.
 * All types namespaced under Address for intuitive access.
 *
 * @example
 * ```typescript
 * import { Address } from './address.js';
 *
 * // Create Address
 * const addr: Address = Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3");
 *
 * // Operations with this: pattern
 * const hex = Address.toHex.call(addr);
 * const checksummed = Address.toChecksumHex.call(addr);
 * ```
 */

import type { Hex } from "./hex.js";
import { Rlp } from "./rlp.js";
import { Hash } from "./hash.js";

// ============================================================================
// Main Address Namespace
// ============================================================================

export namespace Address {
  // ==========================================================================
  // Core Constants
  // ==========================================================================

  export const SIZE = 20;
  export const HEX_SIZE = 42; // 0x + 40 hex chars

  // ==========================================================================
  // Error Types
  // ==========================================================================

  export class InvalidHexFormatError extends Error {
    constructor(message = "Invalid hex format for address") {
      super(message);
      this.name = "InvalidHexFormatError";
    }
  }

  export class InvalidHexStringError extends Error {
    constructor(message = "Invalid hex string") {
      super(message);
      this.name = "InvalidHexStringError";
    }
  }

  export class InvalidAddressLengthError extends Error {
    constructor(message = "Invalid address length") {
      super(message);
      this.name = "InvalidAddressLengthError";
    }
  }

  export class InvalidValueError extends Error {
    constructor(message: string) {
      super(message);
      this.name = "InvalidValueError";
    }
  }

  export class NotImplementedError extends Error {
    constructor(message = "Not implemented") {
      super(message);
      this.name = "NotImplementedError";
    }
  }

  // ==========================================================================
  // Branded Types
  // ==========================================================================

  /**
   * EIP-55 checksummed address hex string
   */
  export type ChecksumHex = Hex & { readonly __checksummed: true };

  // ==========================================================================
  // Universal Constructor
  // ==========================================================================

  /**
   * Create Address from various input types (universal constructor)
   *
   * @param value - Number, bigint, hex string, or Uint8Array
   * @returns Address
   * @throws {InvalidValueError} If value type is unsupported or invalid
   * @throws {InvalidHexFormatError} If hex string is invalid
   * @throws {InvalidAddressLengthError} If bytes length is not 20
   *
   * @example
   * ```typescript
   * const addr1 = Address.from(0x742d35Cc6634C0532925a3b844Bc9e7595f251e3n);
   * const addr2 = Address.from(12345);
   * const _addr3 = Address.from("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3");
   * const addr4 = Address.from(new Uint8Array(20));
   * ```
   */
  export function from(value: number | bigint | string | Uint8Array): Address {
    // Route to appropriate from* method based on type
    if (typeof value === "number" || typeof value === "bigint") {
      return fromNumber(value);
    } else if (typeof value === "string") {
      return fromHex(value);
    } else if (value instanceof Uint8Array) {
      return fromBytes(value);
    } else {
      throw new InvalidValueError("Unsupported address value type");
    }
  }

  // ==========================================================================
  // Conversion Operations
  // ==========================================================================

  /**
   * Parse hex string to Address (standard form)
   *
   * @param hex - Hex string with 0x prefix
   * @returns Address bytes
   * @throws If invalid format or length
   *
   * @example
   * ```typescript
   * const addr = Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3");
   * ```
   */
  export function fromHex(hex: string): Address {
    if (!hex.startsWith("0x") || hex.length !== HEX_SIZE) {
      throw new InvalidHexFormatError();
    }
    // Validate hex characters
    const hexPart = hex.slice(2);
    if (!/^[0-9a-fA-F]{40}$/.test(hexPart)) {
      throw new InvalidHexStringError();
    }
    const bytes = new Uint8Array(SIZE);
    for (let i = 0; i < SIZE; i++) {
      const byte = Number.parseInt(hexPart.slice(i * 2, i * 2 + 2), 16);
      bytes[i] = byte;
    }
    return bytes as Address;
  }

  /**
   * Create Address from raw bytes (standard form)
   *
   * @param bytes - Raw 20-byte array
   * @returns Address
   * @throws If length is not 20 bytes
   *
   * @example
   * ```typescript
   * const bytes = new Uint8Array(20);
   * const addr = Address.fromBytes(bytes);
   * ```
   */
  export function fromBytes(bytes: Uint8Array): Address {
    if (bytes.length !== SIZE) {
      throw new InvalidAddressLengthError();
    }
    return new Uint8Array(bytes) as Address;
  }

  /**
   * Create Address from number value (takes lower 160 bits) (standard form)
   *
   * @param value - Number or bigint value
   * @returns Address from lower 160 bits
   * @throws {InvalidValueError} If value is negative
   *
   * @example
   * ```typescript
   * const addr = Address.fromNumber(0x742d35Cc6634C0532925a3b844Bc9e7595f251e3n);
   * const addr2 = Address.fromNumber(12345);
   * ```
   */
  export function fromNumber(value: bigint | number): Address {
    // Convert number to bigint if needed
    const bigintValue = typeof value === "number" ? BigInt(value) : value;

    // Validate non-negative
    if (bigintValue < 0n) {
      throw new InvalidValueError("Address value cannot be negative");
    }

    const bytes = new Uint8Array(SIZE);
    let v = bigintValue & ((1n << 160n) - 1n);
    for (let i = 19; i >= 0; i--) {
      bytes[i] = Number(v & 0xffn);
      v >>= 8n;
    }
    return bytes as Address;
  }

  /**
   * Create Address from secp256k1 public key (standard form)
   *
   * @param x - Public key x coordinate
   * @param y - Public key y coordinate
   * @returns Address derived from keccak256(pubkey)[12:32]
   *
   * @example
   * ```typescript
   * const addr = Address.fromPublicKey(xCoord, yCoord);
   * ```
   */
  export function fromPublicKey(x: bigint, y: bigint): Address {
    // Encode uncompressed public key (64 bytes: 32 bytes x + 32 bytes y)
    const pubkey = new Uint8Array(64);
    for (let i = 31; i >= 0; i--) {
      pubkey[31 - i] = Number((x >> BigInt(i * 8)) & 0xffn);
      pubkey[63 - i] = Number((y >> BigInt(i * 8)) & 0xffn);
    }
    // Address = keccak256(pubkey)[12:32]
    const hash = Hash.keccak256(pubkey) as unknown as Uint8Array;
    return hash.slice(12, 32) as Address;
  }

  /**
   * Convert Address to hex string
   *
   * @returns Lowercase hex string with 0x prefix
   *
   * @example
   * ```typescript
   * const hex = Address.toHex.call(addr);
   * // "0x742d35cc6634c0532925a3b844bc9e7595f251e3"
   * ```
   */
  export function toHex(this: Address): Hex {
    return `0x${Array.from(this, (b) => b.toString(16).padStart(2, "0")).join("")}` as Hex;
  }

  /**
   * Convert Address to EIP-55 checksummed hex string
   *
   * @returns Checksummed hex string with mixed case
   *
   * @example
   * ```typescript
   * const checksummed = Address.toChecksumHex.call(addr);
   * // "0x742d35Cc6634C0532925a3b844Bc9e7595f251e3"
   * ```
   */
  export function toChecksumHex(this: Address): ChecksumHex {
    const lower = toHex.call(this).slice(2);
    const hashBytes = Hash.keccak256(new TextEncoder().encode(lower)) as unknown as Uint8Array;
    const hashHex = Array.from(hashBytes, (b) => b.toString(16).padStart(2, "0")).join("");
    let result = "0x";
    for (let i = 0; i < 40; i++) {
      const ch = lower[i];
      if (ch !== undefined && ch >= "a" && ch <= "f") {
        const hv = Number.parseInt(hashHex[i] ?? "0", 16);
        result += hv >= 8 ? ch.toUpperCase() : ch;
      } else {
        result += ch ?? "";
      }
    }
    return result as ChecksumHex;
  }

  /**
   * Convert Address to uint256
   *
   * @returns Bigint representation
   *
   * @example
   * ```typescript
   * const value = Address.toU256.call(addr);
   * ```
   */
  export function toU256(this: Address): bigint {
    let result = 0n;
    for (let i = 0; i < SIZE; i++) {
      result = (result << 8n) | BigInt(this[i] ?? 0);
    }
    return result;
  }

  /**
   * Convert Address to ABI-encoded bytes (32 bytes, left-padded)
   *
   * Ethereum ABI encoding pads addresses to 32 bytes by prepending 12 zero bytes.
   *
   * @param this - Address to encode
   * @returns 32-byte ABI-encoded Uint8Array
   *
   * @example
   * ```typescript
   * const addr = Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3");
   * const encoded = Address.toAbiEncoded.call(addr);
   * // encoded.length === 32
   * ```
   */
  export function toAbiEncoded(this: Address): Uint8Array {
    const result = new Uint8Array(32);
    result.set(this, 12); // Left-pad with 12 zero bytes
    return result;
  }

  /**
   * Decode Address from ABI-encoded bytes (32 bytes)
   *
   * Extracts the last 20 bytes from 32-byte ABI-encoded address data.
   *
   * @param bytes - 32-byte ABI-encoded data
   * @returns Decoded Address
   * @throws Error if bytes length is not 32
   *
   * @example
   * ```typescript
   * const encoded = new Uint8Array(32);
   * // ... set encoded[12:32] to address bytes ...
   * const addr = Address.fromAbiEncoded(encoded);
   * ```
   */
  export function fromAbiEncoded(bytes: Uint8Array): Address {
    if (bytes.length !== 32) {
      throw new Error(`ABI-encoded Address must be exactly 32 bytes, got ${bytes.length}`);
    }
    return bytes.slice(12, 32) as Address;
  }

  // ==========================================================================
  // Validation Operations
  // ==========================================================================

  /**
   * Check if address is zero address
   *
   * @returns True if all bytes are zero
   *
   * @example
   * ```typescript
   * if (Address.isZero.call(addr)) {
   *   console.log("Zero address");
   * }
   * ```
   */
  export function isZero(this: Address): boolean {
    return this.every((b) => b === 0);
  }

  /**
   * Check if two addresses are equal
   *
   * @param other - Address to compare with
   * @returns True if addresses are identical
   *
   * @example
   * ```typescript
   * if (Address.equals.call(addr1, addr2)) {
   *   console.log("Addresses match");
   * }
   * ```
   */
  export function equals(this: Address, other: Address): boolean {
    if (this.length !== other.length) return false;
    for (let i = 0; i < this.length; i++) {
      if (this[i] !== other[i]) return false;
    }
    return true;
  }

  /**
   * Check if string is valid address format (standard form)
   *
   * @param str - String to validate
   * @returns True if valid hex format (with or without 0x)
   *
   * @example
   * ```typescript
   * if (Address.isValid("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3")) {
   *   const addr = Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3");
   * }
   * ```
   */
  export function isValid(str: string): boolean {
    if (!str.startsWith("0x")) {
      return str.length === 40 && /^[0-9a-fA-F]{40}$/.test(str);
    }
    return str.length === HEX_SIZE && /^0x[0-9a-fA-F]{40}$/.test(str);
  }

  /**
   * Check if string has valid EIP-55 checksum (standard form)
   *
   * @param str - Address string to validate
   * @returns True if checksum is valid
   *
   * @example
   * ```typescript
   * if (Address.isValidChecksum("0x742d35Cc6634C0532925a3b844Bc9e7595f251e3")) {
   *   console.log("Valid checksum");
   * }
   * ```
   */
  export function isValidChecksum(str: string): boolean {
    if (!isValid(str)) return false;
    try {
      const addr = fromHex(str.startsWith("0x") ? str : `0x${str}`);
      const checksummed = toChecksumHex.call(addr) as string;
      return checksummed === (str.startsWith("0x") ? str : `0x${str}`);
    } catch {
      return false;
    }
  }

  /**
   * Type guard for Address (standard form)
   *
   * @param value - Value to check
   * @returns True if value is an Address
   *
   * @example
   * ```typescript
   * if (Address.is(value)) {
   *   const hex = Address.toHex(value);
   * }
   * ```
   */
  export function is(value: unknown): value is Address {
    return value instanceof Uint8Array && value.length === SIZE;
  }

  // ==========================================================================
  // Special Address Operations
  // ==========================================================================

  /**
   * Create zero address (standard form)
   *
   * @returns Zero address (0x0000...0000)
   *
   * @example
   * ```typescript
   * const zero = Address.zero();
   * ```
   */
  export function zero(): Address {
    return new Uint8Array(SIZE) as Address;
  }

  // ==========================================================================
  // Contract Address Calculation
  // ==========================================================================

  /**
   * Calculate CREATE contract address
   *
   * address = keccak256(rlp([sender, nonce]))[12:32]
   *
   * @param nonce - Transaction nonce
   * @returns Calculated contract address
   * @throws {InvalidValueError} If nonce is negative
   *
   * @example
   * ```typescript
   * const contractAddr = Address.calculateCreateAddress.call(deployerAddr, 5n);
   * ```
   */
  export function calculateCreateAddress(this: Address, nonce: bigint): Address {
    // Validate non-negative nonce
    if (nonce < 0n) {
      throw new InvalidValueError("Nonce cannot be negative");
    }

    // Convert nonce to big-endian bytes and strip leading zeros
    let nonceBytes: Uint8Array;
    if (nonce === 0n) {
      // Special case: nonce 0 encodes as empty bytes for RLP
      nonceBytes = new Uint8Array(0);
    } else {
      // Find minimum bytes needed to represent nonce
      const hex = nonce.toString(16);
      const hexPadded = hex.length % 2 === 0 ? hex : `0${hex}`;
      const byteLength = hexPadded.length / 2;
      nonceBytes = new Uint8Array(byteLength);
      for (let i = 0; i < byteLength; i++) {
        nonceBytes[i] = Number.parseInt(hexPadded.slice(i * 2, i * 2 + 2), 16);
      }
    }

    // RLP encode [sender_address, nonce]
    const encoded = Rlp.encode.call([this, nonceBytes]);

    // Hash the encoded data
    const hash = Hash.keccak256(encoded) as unknown as Uint8Array;

    // Return bytes 12-32 (last 20 bytes) as Address
    return hash.slice(12, 32) as Address;
  }

  /**
   * Calculate CREATE2 contract address
   *
   * address = keccak256(0xff ++ sender ++ salt ++ keccak256(initCode))[12:32]
   *
   * @param salt - 32-byte salt
   * @param initCode - Contract initialization code
   * @returns Calculated contract address
   *
   * @example
   * ```typescript
   * const contractAddr = Address.calculateCreate2Address.call(
   *   deployerAddr,
   *   saltBytes,
   *   initCode
   * );
   * ```
   */
  export function calculateCreate2Address(
    this: Address,
    salt: Uint8Array,
    initCode: Uint8Array,
  ): Address {
    if (salt.length !== 32) {
      throw new Error("Salt must be 32 bytes");
    }

    // address = keccak256(0xff ++ sender ++ salt ++ keccak256(initCode))[12:32]
    const initCodeHash = Hash.keccak256(initCode) as unknown as Uint8Array;
    const data = new Uint8Array(1 + SIZE + 32 + 32);
    data[0] = 0xff;
    data.set(this, 1);
    data.set(salt, 1 + SIZE);
    data.set(initCodeHash, 1 + SIZE + 32);

    const hash = Hash.keccak256(data) as unknown as Uint8Array;
    return hash.slice(12, 32) as Address;
  }

  // ==========================================================================
  // Comparison Operations
  // ==========================================================================

  /**
   * Compare two addresses lexicographically
   *
   * @param other - Address to compare with
   * @returns -1 if this < other, 0 if equal, 1 if this > other
   *
   * @example
   * ```typescript
   * const sorted = addresses.sort((a, b) => Address.compare.call(a, b));
   * ```
   */
  export function compare(this: Address, other: Address): number {
    for (let i = 0; i < SIZE; i++) {
      const thisByte = this[i] ?? 0;
      const otherByte = other[i] ?? 0;
      if (thisByte < otherByte) return -1;
      if (thisByte > otherByte) return 1;
    }
    return 0;
  }

  /**
   * Check if this address is less than other
   *
   * @param other - Address to compare with
   * @returns True if this < other
   */
  export function lessThan(this: Address, other: Address): boolean {
    return compare.call(this, other) < 0;
  }

  /**
   * Check if this address is greater than other
   *
   * @param other - Address to compare with
   * @returns True if this > other
   */
  export function greaterThan(this: Address, other: Address): boolean {
    return compare.call(this, other) > 0;
  }

  // ==========================================================================
  // Formatting Operations
  // ==========================================================================

  /**
   * Format address with shortened display
   *
   * @param prefixLength - Number of chars to show at start (default: 6)
   * @param suffixLength - Number of chars to show at end (default: 4)
   * @returns Shortened address like "0x742d...51e3"
   *
   * @example
   * ```typescript
   * const short = Address.toShortHex.call(addr);
   * // "0x742d...51e3"
   * const custom = Address.toShortHex.call(addr, 8, 6);
   * // "0x742d35...251e3"
   * ```
   */
  export function toShortHex(this: Address, prefixLength?: number, suffixLength?: number): string {
    const prefix = prefixLength ?? 6;
    const suffix = suffixLength ?? 4;

    const hex = toHex.call(this);
    if (prefix + suffix >= 40) return hex;
    return `${hex.slice(0, 2 + prefix)}...${hex.slice(-suffix)}`;
  }

  /**
   * Format address for display (checksummed)
   *
   * @returns Checksummed hex string
   *
   * @example
   * ```typescript
   * console.log(Address.format.call(addr));
   * // "0x742d35Cc6634C0532925a3b844Bc9e7595f251e3"
   * ```
   */
  export function format(this: Address): string {
    return toChecksumHex.call(this);
  }
}

/**
 * Address type - 20-byte Ethereum address
 *
 * Uses TypeScript declaration merging - Address is both a namespace and a type.
 */
export type Address = Uint8Array & { readonly __tag: "Address" };

// Re-export namespace as default
export default Address;

File: primitives/address.go
package primitives

import (
	"errors"
	"fmt"
	"math/big"
	"strings"

	"golang.org/x/crypto/sha3"
)

// Address represents an Ethereum address (20 bytes).
type Address [20]byte

var (
	// ZeroAddress is the zero address (0x0000...0000)
	ZeroAddress = Address{}

	// Address errors
	ErrInvalidAddressLength = errors.New("invalid address length: must be 20 bytes")
	ErrInvalidAddressFormat = errors.New("invalid address format")
)

// FromHex creates an Address from a hex string (with 0x prefix).
// The hex string must be exactly 42 characters (0x + 40 hex digits).
func AddressFromHex(hexStr string) (Address, error) {
	if len(hexStr) != 42 || hexStr[0:2] != "0x" {
		return Address{}, ErrInvalidAddressFormat
	}

	bytes, err := HexToBytes(hexStr)
	if err != nil {
		return Address{}, err
	}

	if len(bytes) != 20 {
		return Address{}, ErrInvalidAddressLength
	}

	var addr Address
	copy(addr[:], bytes)
	return addr, nil
}

// FromBytes creates an Address from a byte slice.
func AddressFromBytes(bytes []byte) (Address, error) {
	if len(bytes) != 20 {
		return Address{}, ErrInvalidAddressLength
	}

	var addr Address
	copy(addr[:], bytes)
	return addr, nil
}

// FromPublicKey derives an Address from an uncompressed public key (x, y coordinates).
func AddressFromPublicKey(publicKeyX, publicKeyY *big.Int) Address {
	// Concatenate x and y coordinates (64 bytes total)
	pubKeyBytes := make([]byte, 64)

	// Write x coordinate (big-endian)
	xBytes := publicKeyX.Bytes()
	copy(pubKeyBytes[32-len(xBytes):32], xBytes)

	// Write y coordinate (big-endian)
	yBytes := publicKeyY.Bytes()
	copy(pubKeyBytes[64-len(yBytes):64], yBytes)

	// Hash and take last 20 bytes
	hash := Keccak256(pubKeyBytes)
	var addr Address
	copy(addr[:], hash[12:32])

	return addr
}

// FromU256 creates an Address from a U256 value (uses last 20 bytes).
func AddressFromU256(value *big.Int) Address {
	var addr Address
	bytes := value.Bytes()

	// If value is less than 20 bytes, pad with leading zeros
	if len(bytes) <= 20 {
		copy(addr[20-len(bytes):], bytes)
	} else {
		// If value is more than 20 bytes, take the last 20 bytes
		copy(addr[:], bytes[len(bytes)-20:])
	}

	return addr
}

// ToHex converts the Address to a hex string (lowercase, with 0x prefix).
func (a Address) ToHex() string {
	return BytesToHex(a[:])
}

// ToChecksumHex converts the Address to an EIP-55 checksummed hex string.
func (a Address) ToChecksumHex() string {
	// Convert to lowercase hex without prefix
	hex := strings.ToLower(fmt.Sprintf("%x", a[:]))

	// Hash the lowercase hex
	hash := sha3.NewLegacyKeccak256()
	hash.Write([]byte(hex))
	hashBytes := hash.Sum(nil)

	// Build checksummed string
	result := make([]byte, 42)
	result[0] = '0'
	result[1] = 'x'

	for i := 0; i < 40; i++ {
		char := hex[i]
		if char >= 'a' && char <= 'f' {
			// Check if this hex digit should be uppercase
			hashByte := hashBytes[i/2]
			var nibble byte
			if i%2 == 0 {
				nibble = hashByte >> 4
			} else {
				nibble = hashByte & 0x0f
			}

			if nibble >= 8 {
				char = char - 'a' + 'A'
			}
		}
		result[2+i] = char
	}

	return string(result)
}

// IsZero returns true if the address is the zero address.
func (a Address) IsZero() bool {
	return a == ZeroAddress
}

// Equals returns true if two addresses are equal.
func (a Address) Equals(other Address) bool {
	return a == other
}

// ToU256 converts the Address to a U256 (big.Int).
func (a Address) ToU256() *big.Int {
	return new(big.Int).SetBytes(a[:])
}

// Bytes returns the address as a byte slice.
func (a Address) Bytes() []byte {
	return a[:]
}

// String returns the checksummed hex representation of the address.
func (a Address) String() string {
	return a.ToChecksumHex()
}

// ValidateChecksum validates an EIP-55 checksummed address.
func ValidateChecksum(hexStr string) bool {
	addr, err := AddressFromHex(hexStr)
	if err != nil {
		return false
	}

	return addr.ToChecksumHex() == hexStr
}

// IsValidAddress checks if a string is a valid address format (hex validation only).
func IsValidAddress(addrStr string) bool {
	if len(addrStr) != 42 || addrStr[0:2] != "0x" {
		return false
	}

	for i := 2; i < len(addrStr); i++ {
		c := addrStr[i]
		valid := (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
		if !valid {
			return false
		}
	}

	return true
}

// CalculateCreateAddress calculates the CREATE contract address.
// addr = keccak256(rlp([sender, nonce]))[12:]
func CalculateCreateAddress(sender Address, nonce uint64) (Address, error) {
	// Simple RLP encoding for [address, nonce]
	// This is a simplified implementation - in production, use a full RLP library
	rlpEncoded := encodeCreateRLP(sender[:], nonce)

	hash := Keccak256(rlpEncoded)
	var addr Address
	copy(addr[:], hash[12:32])

	return addr, nil
}

// CalculateCreate2Address calculates the CREATE2 contract address.
// addr = keccak256(0xff ++ sender ++ salt ++ keccak256(initCode))[12:]
func CalculateCreate2Address(sender Address, salt [32]byte, initCodeHash [32]byte) Address {
	// Build: 0xff ++ sender (20 bytes) ++ salt (32 bytes) ++ initCodeHash (32 bytes)
	data := make([]byte, 1+20+32+32)
	data[0] = 0xff
	copy(data[1:21], sender[:])
	copy(data[21:53], salt[:])
	copy(data[53:85], initCodeHash[:])

	hash := Keccak256(data)
	var addr Address
	copy(addr[:], hash[12:32])

	return addr
}

// Keccak256 computes the Keccak-256 hash of data.
func Keccak256(data []byte) []byte {
	hash := sha3.NewLegacyKeccak256()
	hash.Write(data)
	return hash.Sum(nil)
}

// encodeCreateRLP is a simplified RLP encoding for CREATE address calculation.
// This encodes [address_bytes, nonce] according to RLP rules.
func encodeCreateRLP(address []byte, nonce uint64) []byte {
	// Encode address (20 bytes) - string encoding
	addressRLP := encodeRLPBytes(address)

	// Encode nonce - integer encoding
	var nonceRLP []byte
	if nonce == 0 {
		nonceRLP = []byte{0x80} // Empty byte array
	} else if nonce < 0x80 {
		nonceRLP = []byte{byte(nonce)}
	} else {
		// Encode as byte array (minimal bytes)
		nonceBytes := encodeMinimalBytes(nonce)
		nonceRLP = encodeRLPBytes(nonceBytes)
	}

	// Combine into list
	payload := append(addressRLP, nonceRLP...)
	return encodeRLPList(payload)
}

// encodeRLPBytes encodes bytes according to RLP rules.
func encodeRLPBytes(data []byte) []byte {
	if len(data) == 1 && data[0] < 0x80 {
		return data
	}
	if len(data) <= 55 {
		return append([]byte{byte(0x80 + len(data))}, data...)
	}
	lenBytes := encodeMinimalBytes(uint64(len(data)))
	prefix := byte(0xb7 + len(lenBytes))
	result := append([]byte{prefix}, lenBytes...)
	return append(result, data...)
}

// encodeRLPList encodes a list payload according to RLP rules.
func encodeRLPList(payload []byte) []byte {
	if len(payload) <= 55 {
		return append([]byte{byte(0xc0 + len(payload))}, payload...)
	}
	lenBytes := encodeMinimalBytes(uint64(len(payload)))
	prefix := byte(0xf7 + len(lenBytes))
	result := append([]byte{prefix}, lenBytes...)
	return append(result, payload...)
}

// encodeMinimalBytes encodes a uint64 as minimal big-endian bytes.
func encodeMinimalBytes(value uint64) []byte {
	if value == 0 {
		return []byte{}
	}

	// Determine how many bytes we need
	var buf [8]byte
	var i int
	for i = 7; i >= 0; i-- {
		buf[i] = byte(value & 0xff)
		value >>= 8
		if value == 0 {
			break
		}
	}

	return buf[i:]
}

// constantTimeCompare performs constant-time comparison of byte slices.
func constantTimeCompare(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	var result byte
	for i := 0; i < len(a); i++ {
		result |= a[i] ^ b[i]
	}
	return result == 0
}
